### ISSUE TITLE
Suggestion: sg_update_image to take in an x, y offset to allow partial texture update

### ISSUE THREAD

**Issue by oviano** [floooh/sokol#567]
I am trying to implement a dynamic texture atlas and I realised that at the moment there doesn't seem to be a way of only updating a certain region of a texture.

Is it as simple as being able to pass (x, y, w, h) into sg_update_image() and have this propagated down to the relevant Metal replaceRegion, GL glTexSubImage2D etc, the assumption being that the data is structured in the same way as the original data (bytes per row etc)?

**Comment by floooh** [comment]
I have stuff like this planned for a complete overhaul of the resource update functions, but it will be a while until I'll get around to this. There might be some destination-coordinates restrictions in the different 3D APIs which I'm not yet aware of (e.g. I could imagine that updating compressed images only works at block granularity).

But a bigger problem is the restriction of only one update per frame, so you can't simply build a texture from different source rectangle in the same frame, also since dynamic textures are internally double-buffered, partial updates in two follupup frames would actually update different textures.

For a current workaround see sokol_fontstash.h, this keeps a backup "texture" (just a memory blob) in system RAM which is managed and updated by FontStash, and FontStash sets a dirty-flag when the sokol-texture needs be updated, which then copies the whole texture data: 

https://github.com/floooh/sokol/blob/896497ad775153ebe1385c9f95d296860ef53714/util/sokol_fontstash.h#L1768-L1779

Eventually I want to create a "proper" set of dynamic resource update functions which allow to copy regions of buffers and textures between cpu- and gpu-memory, and also between gpu resources.

**Comment by floooh** [comment]
Btw, I just found this in PVRTC texture compression spec:

```
    5) Is sub-texturing supported?

       Resolution: Only for the reloading of complete
       images. Sub-images are not supportable because the PVRTC
       algorithm uses significant adjacency information, so there is
       no discrete block of texels that can be decoded as a standalone
       sub-unit, and so it follows that no stand alone sub-unit of
       data can be loaded without changing the decoding of surrounding
       texels.   
```

...so at least for compressed texture formats, and at least PVRTC-compressed textures, partial image updates won't be possible.

**Comment by leiradel** [comment]
+1

My use case is exactly the same as the OP. Being able to update parts of a texture without having to hold a copy of the pixels would be great.

**Comment by leiradel** [comment]
I know this is not helpful, but as I'm coding, I've found another situation for partial texture updates.

I'm emulating consoles and computers, and some can change their resolution as needed by the game. I know the maximum width and height, and could allocate a texture that big and update only part of the sub-texture depending on the current resolution.

Right now, I have to always destroy and re-create the texture when the resolution changes.

**Comment by darkuranium** [comment]
My own use-case is virtual texturing (not necessarily with sparse textures). I want to essentially be able to fit large amounts of data into fixed amounts of RAM.

It *might* be a common enough use-case to warrant a separate extension (as a stop-gap without needing "full-blown" partial updates support --- though such an approach could take advantage of sparse texturing extensions).

**Comment by kkukshtel** [comment]
Alos +1-ing this — exactly for a font texture atlas use.

**Comment by artheru** [comment]
a coarse but working snippet:

`
void updateTextureW4K(sg_image simg, int objmetah, const void* data, sg_pixel_format format)
{
	_sg_image_t* img = _sg_lookup_image(&_sg.pools, simg.id);

	_sg_gl_cache_store_texture_binding(0);
	_sg_gl_cache_bind_texture(0, img->gl.target, img->gl.tex[img->cmn.active_slot]);
	GLenum gl_img_target = img->gl.target;
	glTexSubImage2D(gl_img_target, 0,
		0, 0,
		4096, objmetah,
		_sg_gl_teximage_format(format), _sg_gl_teximage_type(format),
		data);
	_sg_gl_cache_restore_texture_binding(0);
}
`

**Comment by darkuranium** [comment]
I did some thinking on this, and one way to get this done in a way that's compatible with different APIs would be to introduce explicit sync points into the sokol_gfx API.

Not the best solution, but it *would* open up other possibilities in the future (e.g. compute shaders). Essentially something like:
```c
sg_sync_point sync;
sg_update_subimage(img, &sync, data);
...
sg_sync(&sync); // or `sg_barrier` or whatever you call it
sg_draw(...);
```

(in things like OpenGL [ES], where `glTexSubImage{2D,...}` are auto-synced, `sg_sync` is simply a no-op &hellip; it would be nice to be able to make a small buffer write dependent upon another sync point, exactly for the purpose of virtual texturing, though)

I'd argue it'd fit a "sokol_gfx V2" more than the existing API, though; for one, it's a lower-level concept, and a completely new object/API concept. Arguably also not 100% in line with Sokol's philosophy (there's a reason as to why sokol_gfx does ring buffers for you, after all), but.

The alternative to explicit syncing is to simply do this internally, at the cost of stalls whenever you do a draw that reads from the same texture; an acceptable cost in many cases, but it causes problems for some use-cases[^1]. It *would* be closer to sokol's current APIs though.

[^1]: Basically wherever you might want to keep rendering despite parts of the texture not being updated yet &mdash; a typical case would be virtual textures, but also possibly font atlases, if you accept that some characters might be invisible for a few frames.

**Comment by randrew** [comment]
> ...so for compressed texture formats (at least PVRTC-compressed textures), partial image updates won't be possible.

Well, PVRTC is gone now :)

> But a bigger problem is the restriction of only one update per frame, so you can't simply incrementally build a texture in the same frame, But even when distributing the update over several frames, the next problem is that dynamic textures are internally double-buffered, partial updates in two follupup frames would actually update different textures.

Without requiring a big overhaul that allows calling sg_update_image() multiple times per frame, what about something like,

```c
sg_partial_group_t group;
sg_partial_piece_t partials[32];
sg_begin_partial_update(&group, &my_image, ...);
for (int i = 0; i < pieces_this_frame; i++) {
	...
	sg_add_partial_piece(&group, &partials[i], &subresource, x, y, width, height, ...);
}
sg_finish_partial_update(&group);
```

**Comment by jiangbo** [comment]
Can I update a single layer in a texture array? Currently I have to wait for all the images to load before I can generate the texture array. Imagine I have 10 textures at 4096×4096; I have to fully decompress them in memory and then update them all at once. If I could update them individually, my peak memory usage would definitely be much lower. How can I solve this problem? Thanks

**Comment by floooh** [comment]
> Can I update a single layer in a texture array?

Not via the update function, the only way currently is to use a render- or compute-pass to get the data into individual texture layers.

The only other alternative is to directly call 3D backend API functions, you can get the underlying backend 3D resources via `sg_[backend]_[resourcetype]_info()` functions (e.g. `sg_d3d11_image_info()` to get the D3D11 texture object pointers.

Or alternatively you can also create the layer texture outside sokol_gfx.h (via direct calls into the backend API) and then 'inject' the resource into sokol_gfx.h (e.g. see the platform-specific inject samples):

- https://github.com/floooh/sokol-samples/blob/master/d3d11/inject-d3d11.c
- https://github.com/floooh/sokol-samples/blob/master/glfw/inject-glfw.c
- https://github.com/floooh/sokol-samples/blob/master/html5/inject-emsc.c
- https://github.com/floooh/sokol-samples/blob/master/metal/inject-metal.m

...the experimental Vulkan backend currently doesn't support injecting native resources.

**Comment by jiangbo** [comment]
Dealing with all the different platforms is too complicated (ㄒoㄒ). I'm not in a rush to need this at the moment, though—I'm just asking to see if this can actually be implemented across various platforms? And does Sokol have any plans to implement this?

**Comment by floooh** [comment]
> And does Sokol have any plans to implement this?

Yes definitely, there will be a complete overhaul of the resource update functions, roughly similar to what WebGPU provides (e.g. more flexible CPU => GPU, GPU => GPU and maybe also GPU => CPU copies, while also getting rid of the restriction to do only one update per resource and frame). The only downside will be that the implementation for OpenGL may be non-optimal (GL is the main reason why the current update API is so restricted, since it's too easy in GL to introduce lock-stalls). No ETA yet, but it'll be one of the next "big" features.

**Comment by darkuranium** [comment]
For what it's worth, this feature is important enough to me that I was forced to stop using `sokol_gfx` because of it.
(there was also a second missing feature, but I managed to work around it [^1])

[^1]: I needed the ability to *read* texture data. The rationale of avoiding things that hurt performance is sensible, but in my case, I needed it for an in-game screenshot feature, and people are already used to stutters while taking screenshots *anyway*. I did manage to work around it by hacking the `sokol_gfx` source to introduce it for the APIs I use.

TL;DR I'm implementing virtual texturing, where I keep a cache of tiles in use and page them in &amp; out. This usually means huge textures (I typically use a `texture2DArray` of 16384×8192×N on desktop; less on mobile, but same idea), so simply doing a full update each time is far too wasteful in bandwidth.

More generally, it applies to any form of dynamic texture atlases where the atlas is large enough and/or updated commonly enough that a full upload is prohibitive.

---

As for my $0.02 w.r.t. implementation (*see the last code block for a TL;DR*):

One approach that I liked and I think might work for you is the SDLGPU3 approach of making the cycling explicit &mdash; essentially the *user* specifies which target textures can be cycled at the start of each render pass.

In terms of `sokol_gfx`, this means dropping the "implicit resource cycling" feature, and instead adding a `sg_cycle_mode cycle` field to each target texture/view in `sg_begin_pass()`.

Something like this:

```c
/*
    Default could be (YMMV):
    - KEEP (different from old behavior but probably a saner default with the new API due to --- but a performance pitfall)
    - CYCLE (same as old behaviur, but probably not the best default in an API with excplicit cycle control --- maintains performance, but it's a programming/usage pitfall)
    - no default (must be explicitly specified --- it's an error not to do so)

    Addendum: Naturally, `CYCLE` can be implemented in terms of `FORCE_CYCLE` in the short term, and only separated later.
*/
typedef enum sg_cycle_mode {
    _SG_CYCLEMODE_DEFAULT,
    SG_CYCLEMODE_KEEP,   // don't cycle the texture, even if it means a stall
    SG_CYCLEMODE_CYCLE, // cycle the texture depending on heuristics (this is the old behavior, I think?) [could also be named "AUTO"?]
    SG_CYCLEMODE_FORCE_CYCLE,  // cycle the texture, always (not 100% on where it'd be useful as of yet, but it feels like there are use-cases out there that would need this --- maybe involving compute vs render pass interactions?)
    _SG_CYCLEMODE_FORCE_U32 = 0x7FFFFFFF
} sg_cycle_mode;

typedef struct sg_view_attachment {
  sg_view view;
  sg_cycle_mode cycle;   // a simple `bool` would work here too (for KEEP vs FORCE_CYCLE)
} sg_view_attachment;

typedef struct sg_attachments {
  sg_view_attachment colors[SG_MAX_COLOR_ATTACHMENTS];
  sg_view resolves[SG_MAX_COLOR_ATTACHMENTS]; // (not sure if this one also needs cycling control?)
  sg_view_attachment depth_stencil;
} sg_attachments;
```

### Short-term:

What I think you could do ***in the short term*** (before a full API rework), in a backwards-compatible way (API-wise, not ABI, though I don't think the latter matters here?) is to modify `sg_attachments` by simply *adding* fields instead of introducing `sg_view_attachment`:

```c
typedef struct sg_attachments {
    sg_view colors[SG_MAX_COLOR_ATTACHMENTS];
    sg_cycle_mode cycle_colors[SG_MAX_COLOR_ATTACHMENTS];  //< ADDED
    sg_view resolves[SG_MAX_COLOR_ATTACHMENTS];
    sg_view depth_stencil;
    sg_cycle_mode cycle_depth_stencil;  //< ADDED
} sg_attachments;
```

Since the changes are purely additive, they should be 100% compatible with old code that does `{ .colors[0] = ... }` and such.

In this case, it would pretty much be 100% necessary to make the default for `sg_cycle_mode` be `SG_CYCLEMODE_CYCLE`; to avoid footguns when using older code with this newer API (since we can't leverage the compiler to issue errors about missing `.cycle_colors[i]` assignments, unlike with a `.colors[i]` type change &mdash; which means the default behavior must match the old behavior). It also *needs* to be an `enum` here instead of `bool` because the values will default to `0`.

This approach could also work in the long term, of course; it's a question of API ergonomics on whether to use a `sg_attachment_view`  or not.

**Comment by floooh** [comment]
I'm not a fan of SDL's manual resource cycling, because it's generally not needed on modern 3D APIs (and also not in some traditional APIs like D3D11).

Instead I will more likely differentiate between 'transient' and 'persistent' updates.

- Transient updates uploads data that's used in the same frame for rendering but won't survive into the next frame, this is similar to the current restriction of "one update per resource per frame", and all data that's used in rendering must be provided. There will be no 'x/y/w/h' placement for transient updates. On some 3D APIs, transient updates won't even have a backing resource, they will only be copied into a global per-frame buffer (which is rotated between frames).
- Persistent updates are blitted into resources and the data persists between frames (e.g. once the data is copied it stays in the resource). Those updates will allow 'placement' into a subregion of the resource. On modern APIs the data will first be copied into a staging buffer, and a command is recorded to blit the staging buffer into the target resource. There may be a pipeline flush involved when the uploaded persistent data is above some tweakable per-frame threshold.

The Vulkan backend is already pretty close to the above resource update strategy, the remaining work is basically to update the public API and 'backport' those update strategies to the other 3D APIs (with various compromises for OpenGL, and I'll may also need to switch to using 'resource heaps' on Metal. This update will then also remove all cases of 'internally double buffered' resources.

PS: this ignores reading back data from the GPU to the CPU, but I think cycling is still not the correct way to do this. Instead it will be a 'reverse blit' from the GPU into a sort-of 'reverse staging buffer' which holds the data until the API user can copy it out into his own memory chunk in the next frame.

**Comment by floooh** [comment]
I'll close this as "nearly done".

The work on the new resource update API is (slowly) underway (see planning ticket here: https://github.com/floooh/sokol/issues/1547), write-unsealed and write-transient has been implemented which allows to write to a subregion of an image.

write-persistent is missing but next on the list.

The new write-functions also allow multiple calls per frame (with the caveat that write-transient and write-persistent calls are only allowed until the written resource is first used in this frame).
